README.md
# connectivity_wall

A Connectivity wall monitor

## Getting Started

Connectivity wall redirect user to a widget when offline and another if online.

We keep the app state using a IndexedStack.



```
import 'package:flutter/material.dart';
import 'package:connectivity_wall/connectivity_wall.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: ConnectivityWall(
          pingInterval: 120,
          responseCode: [200,302,301],
          onPingUrl: Uri.parse("https://pub.dev/"),

          /// Connectedwall
          onConnectedWall: OnlineState(),
          /// User changed from wifi to data or else
          onConnectivityChanged: (result) {
            // ConnectivityResult.mobile
            // ConnectivityResult.wifi
            // ConnectivityResult.none
            print(result);
          },

          /// Disconnected callback
          onDisconnected: () {
            print("Offline, do something");
          },

          /// Disconnected Widget wall
          onDisconnectedWall: OfflineState(),
        ));
  }
}

class OfflineState extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        alignment: Alignment.center,
        child: Center(
          child: Text("Not connected to internet"),
        ),
      ),
    );
  }
}


class OnlineState extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        alignment: Alignment.center,
        child: Center(
          child: Text("Connected "),
        ),
      ),
    );
  }
}


```

CHANGELOG.md
## [0.0.1] - 04/24/21

* First release with example.

Connectivity_wall helps to monitor the internet connectivity and notify users with a  widget wall.

## [0.0.2] - 04/24/21

* Added support to Web platform

## [1.0.1] - 04/29/21

* First major release, after my first like.

## [1.0.2] - 04/30/21

* Added repo to pubspec

## [1.0.5] - 03/30/22

* Upgraded connectivity to connectivity_plus
* Fix initial state

## [1.0.6] - 04/01/22

* Fix static analysis issues

## [1.0.7] - 04/02/22

* Fix static analysis issues

## [1.0.8] - 04/04/22

* Fix lifecycle state: defunct, not mounted

## [1.0.11] - 11/06/2023

* upgrade dependencies:
  * connectivity_plus
  * http
* Upgrade environment:
  * sdk = '>=3.0.3 <4.0.0'


## [1.0.11] - 11/06/2023

* fix Null check:


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:connectivity_wall_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/connectivity_wall_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:connectivity_wall/connectivity_wall.dart';

void main() {
  test('Testes', () {});
}


lib/connectivity_wall.dart
library connectivity_wall;

import 'dart:async';
import 'dart:io';
import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

/// ConnectivityWall widget
class ConnectivityWall extends StatefulWidget {
  /// We make a HEAD request to url every pingInterval
  final Uri onPingUrl;
  ///Interval in secs to check the onPingUrl
  final int pingInterval;

  /// Expect this response code from pingUrl
  final List<int> responseCode;

  /// build this when disconnected
  final Widget onDisconnectedWall;

  /// build this when connected
  final Widget onConnectedWall;

  /// Called on request Timeout
  final Function() onDisconnected;

  /// Return the ping response status code
  final Function(int statusCode) onPingResponse;

  /// Called when user change networks (Wifi,mobile,etc)
  final Function(ConnectivityResult result) onConnectivityChanged;

  /// Constructor
  ConnectivityWall(
      {required this.onPingUrl,
      required this.responseCode,
      required this.onDisconnectedWall,
      required this.onConnectedWall,
      required this.onDisconnected,
      required this.onConnectivityChanged,
      required this.pingInterval,
      required this.onPingResponse});

  @override
  _ConnectivityWallState createState() => _ConnectivityWallState();
}

class _ConnectivityWallState extends State<ConnectivityWall> {
  /// Rebuild based
  bool _isConnected = true;
  Map<bool, int> _index = {false: 0, true: 1};

  StreamSubscription<ConnectivityResult>? subscription;

  Future<http.Response> requestHead(
      Uri url, Map<String, String> requestHeaders) async {
    try {
      final request = await http.Client()
          .head(url, headers: requestHeaders)
          .timeout(const Duration(seconds: 60));
      return request;
    } on TimeoutException catch (e) {
      return http.Response(e.toString(), 408);
    } on SocketException catch (e) {
      return http.Response(e.toString(), 500);
    } on Exception catch (e) {
      return http.Response(e.toString(), 500);
    }
  }

  /// Call widget onConnectivityChanged
  void onChanged(ConnectivityResult result) {
    widget.onConnectivityChanged(result);
    /// Check connection
    ping();
  }

  /// Ping loop, every  pingInterval
  void pingLoop() {
    if (mounted) {
      ping();
      Future.delayed(
          Duration(seconds: widget.pingInterval),
          () => {
                pingLoop(),
              });
    }
  }

  Future<void> ping() async {

    final resposta = await requestHead(widget.onPingUrl, {});
    if (mounted) {
      setState(() {
        _isConnected = (widget.responseCode.contains(resposta.statusCode));
      });
      widget.onPingResponse(resposta.statusCode);
      if(!_isConnected) {
        widget.onDisconnected();
      }


    }
  }

  @override
  void initState() {
    super.initState();
    Connectivity()
        .checkConnectivity()
        .then((result) => {
              /// Call onConnectivityChanged
              onChanged(result),
              /// Subscribe to listen changes
              subscription =
                  Connectivity().onConnectivityChanged.listen(onChanged),
            })
        .then((_) => {
              /// Lets start pinging
              pingLoop(),
            });
  }

  @override
  void dispose() {
    // TODO: implement dispose
    super.dispose();
    subscription?.cancel();
  }

  @override
  Widget build(BuildContext context) {
    return IndexedStack(
      index: _index[_isConnected],
      children: [widget.onDisconnectedWall, widget.onConnectedWall],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:connectivity_wall/connectivity_wall.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          // This is the theme of your application.
          //
          // Try running your application with "flutter run". You'll see the
          // application has a blue toolbar. Then, without quitting the app, try
          // changing the primarySwatch below to Colors.green and then invoke
          // "hot reload" (press "r" in the console where you ran "flutter run",
          // or simply save your changes to "hot reload" in a Flutter IDE).
          // Notice that the counter didn't reset back to zero; the application
          // is not restarted.
          primarySwatch: Colors.blue,
        ),
        home: ConnectivityWall(
          onPingResponse: (response) {},
          pingInterval: 120,
          responseCode: [200],
          onPingUrl: Uri.parse("https://pub.dev/"),

          /// Connectedwall
          onConnectedWall: MyHomePage(title: 'Flutter Demo Home Page'),

          /// User changed from wifi to data or else
          onConnectivityChanged: (result) {
            // ConnectivityResult.mobile
            // ConnectivityResult.wifi
            // ConnectivityResult.none
            print(result);
          },

          /// Disconnected callback
          onDisconnected: () {
            print("Offline, do something");
          },

          /// Disconnected Widget wall
          onDisconnectedWall: OfflineState(),
        ));
  }
}

class OfflineState extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        alignment: Alignment.center,
        child: Center(
          child: Text("Not connected to internet"),
        ),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title!),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
